### PR TITLE
FIX-#4208: Fix lazy metadata update for `PandasDataFrame.from_labels`

### LINKED ISSUE #4208: `modin_df.reset_index` is failed with `ValueError: cannot insert colname, already exists` but pandas works
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **Modin version** (`modin.__version__`): b7027597985a442734a57949d72e3fa0625b4d43
- **Python version**: 3.8.11
- **Code we can use to reproduce**:

```python
# import pandas as pd
import modin.pandas as pd

df = pd.DataFrame({"col0": [0,1,2,3]}, index=pd.Index([11, 22, 33, 44], name="col0"))

df.columns = ["col0_1"]
df = df.reset_index()

print(df)
```

### Describe the problem

Modin throws an error but pandas doesn't. The issue is observed in Santander workload. A possible reason is not correct propagating of new names of columns to lowest levels of architecture.

### Source code / logs

Modin output:
```bash
During handling of the above exception, another exception occurred:

ray::apply_list_of_funcs() (pid=2800927, ip=10.241.129.69)
  File "/localdisk/aprutsko/modin/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py", line 417, in apply_list_of_funcs
    partition = func(partition.copy(), *args, **kwargs)
  File "/localdisk/aprutsko/modin/modin/core/dataframe/pandas/dataframe/dataframe.py", line 827, in from_labels_executor
    return df.reset_index()
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 5841, in reset_index
    new_obj.insert(0, name, level_values)
  File "/localdisk/aprutsko/miniconda/envs/modin/lib/python3.8/site-packages/pandas/core/frame.py", line 4442, in insert
    raise ValueError(f"cannot insert {column}, already exists")
ValueError: cannot insert col0, already exists
```

Pandas output:
```bash
   col0  col0_1
0    11       0
1    22       1
2    33       2
3    44       3
```

### PR DESCRIPTION
**Pull request by prutskov**
Signed-off-by: Alexey Prutskov <alexey.prutskov@intel.com>

## What do these changes do?

This PR adds updating of columns metadata step to call queue before `from_labels` flow.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4208  
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by devin-petersohn — COMMENTED** [review]
Could we do replace the `reset_index` on the partitions with an `insert(..., allow_duplicates=True)` instead of force-propagating the labels all the time? This will be really expensive in the common case (no duplication) and we already error check at the outer layers.

**Comment by prutskov** [comment]
> Could we do replace the `reset_index` on the partitions with an `insert(..., allow_duplicates=True)` instead of force-propagating the labels all the time? This will be really expensive in the common case (no duplication) and we already error check at the outer layers.

@devin-petersohn, I tried to implement the fix via `insert(..., allow_duplicates=True)` in 3de92a7c0fd2838b1c2226cb895468c6ae871ae6. Using this solution we free the common case (without duplication) from additional columns propagating but make a lot of manipulations with index to add each index level to partition-dataframe as column. Does it make sense?

File: modin/core/dataframe/pandas/dataframe/dataframe.py
Comment (devin-petersohn) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:848]: I think we don't need the extra checks, we should just be able to handle all of the cases in the same way here.
```suggestion
            columns_to_insert = df.index.to_frame().reset_index(drop=True)
            df = df.reset_index(drop=True)
            pandas.concat([columns_to_insert, df], axis=1)
```

**Review by devin-petersohn — COMMENTED** [review]
Left a comment, does it make sense?

File: modin/core/dataframe/pandas/dataframe/dataframe.py
Comment (prutskov) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: I think that your suggestion should be under `if any(level_name in df.columns for level_name in df.index.names):` condition, because it will be not efficient to use those number of operations for each `from_labels_executor` call. This idea is implemented in 10c2c99240eb94f90b12bfeea480934b35ee75de.

File: modin/core/dataframe/pandas/dataframe/dataframe.py
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:831]:
```suggestion
                columns_to_add = df.index.to_frame()
                columns_to_add.reset_index(drop=True, inplace=True)
                df = df.reset_index(drop=True)
                return pandas.concat([columns_to_add, df], axis=1, copy=False)
```
should be a bit faster as we don't create unnecessary copies

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:831]: It would also be good to place a comment before the if-branch of why we ignore columns duplication in the kernel function. It's hard to understand the logic of having these two if-branches out of the context of this PR.

Comment (prutskov) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: > should be a bit faster as we don't create unnecessary copies

Applied with additional changes: ` df = df.reset_index(drop=True)` --> `df.reset_index(drop=True, inplace=True)`.

> It would also be good to place a comment before the if-branch of why we ignore columns duplication in the kernel function. It's hard to understand the logic of having these two if-branches out of the context of this PR.

Agree, the note-comment is added.

Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: I feel safe to set `inplace=True` for `columns_to_add` frame as it is a COPY of the `df.index` and so nothing depends on it. I don't feel the same about `df`, since it's the object that is stored in the partition and so modifying it we modify the partition's content, isn't it?

Comment (prutskov) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: Looks reasonable, fixed.

**Review by dchigarev — DISMISSED** [review]
LGTM.

@devin-petersohn had some comments, so waiting for his approval to merge

**Review by devin-petersohn — APPROVED** [review]
>I think that your suggestion should be under if any(level_name in df.columns for level_name in df.index.names): condition, because it will be not efficient to use those number of operations for each from_labels_executor call. This idea is implemented in https://github.com/modin-project/modin/commit/10c2c99240eb94f90b12bfeea480934b35ee75de.

I see what you mean, but internally pandas implements the `reset_index` as a `concat`. This implementation is still ok with me. Thanks for addressing my comments!

**Review by devin-petersohn — APPROVED** [review]

**Pull request merged**
